Update aliasing in tensorexpr fuser

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43743 Update aliasing in tensorexpr fuser**
* #43742 use types in the IR instead of vmap
* #43636 Add inline call to profiling executor pipeline
* #43635 Use prim::TensorExprGroup interned symbol
* #43634 Update requires grad property

Differential Revision: [D23385205](https://our.internmc.facebook.com/intern/diff/D23385205)

Commit:
Update aliasing in tensorexpr fuser

[ghstack-poisoned]

Commit:
Update on "Update aliasing in tensorexpr fuser"




[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -208,12 +208,64 @@ class TensorExprFuser {
+      existing_values.push_back(existing);
+    }
+    std::unordered_map<Value*, Value*> vmap;
+    Node* fusion_group = merge_fn(vmap);
```
There is a single user of this function and for that user `merge_fn` = `SubgraphUtils::mergeNodeIntoSubgraph`. Do we really need to use a lambda here? Feels like we could be over-engineering it here.

eellison replied:
It's also `getOrCreateTensorExprSubgraph`

ZolotukhinM commented on test/jit/test_profiler.py:
```diff
@@ -83,3 +83,20 @@ def test_fuse(a, b):
+        x = torch.ones(1)
+        y = torch.ones(1)
+        foo(x, y)
+        b = foo(x, y)
```
Nit: `b` seems to be unused.

ZolotukhinM commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -208,12 +208,64 @@ class TensorExprFuser {
   }
 
  private:
+  Node* aliasingSafeSubgraphMerge(
```
Might be useful to have a comment explaining what the function does and what's the semantics of its arguments.

ZolotukhinM reviewed (approved):
The `aliasingSafeSubgraphMerge` looks a bit icky to me: I hoped that we could somehow get away without tracking what outputs existed before the merge - it feels to me that we're lacking some basic API and because of that have to leak abstractions.

That said, since it gets the job done, I think we should go with it and consider cleaning it up later.

eellison commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -1379,7 +1379,6 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
             // Check every input's shape against profiled (expected) shape.
             for (i = 0; i < num_inputs; i++) {
               auto& input = peek(stack, i, num_inputs);
-              TORCH_INTERNAL_ASSERT(input.isTensor());
```
this isnt really necessary to do since `toTensor()` has the same assertion - since its a one line change i grouped it in here

eellison commented:
Good comments, i'll try to merge some of my responses into as comments into the code

Commit:
Update on "Update aliasing in tensorexpr fuser"




[ghstack-poisoned]

Commit:
Update on "Update aliasing in tensorexpr fuser"


Differential Revision: [D23385205](https://our.internmc.facebook.com/intern/diff/D23385205)

[ghstack-poisoned]

Commit:
Update on "Update aliasing in tensorexpr fuser"


Differential Revision: [D23385205](https://our.internmc.facebook.com/intern/diff/D23385205)

[ghstack-poisoned]

Commit:
Update on "Update aliasing in tensorexpr fuser"


Differential Revision: [D23385205](https://our.internmc.facebook.com/intern/diff/D23385205)

[ghstack-poisoned]
